src/usr/local/www/interfaces_gre_edit.php

```diff
@@ -55,38 +59,80 @@
+	} else if ($tunnel_type === 'v6') {
+		$reqdfields = explode(" ", "if remote-addr tunnel-local-addr6 tunnel-remote-addr6 tunnel-remote-net6");
+		$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IPv6 address"), gettext("Local tunnel IPv6 address"), gettext("Remote tunnel IPv6 address"), gettext("Remote IPv6 tunnel network"));
+	} else {
+		$reqdfields = explode(" ", "if remote-addr tunnel-local-addr tunnel-remote-addr tunnel-remote-net");
+		$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IPv4 address"), gettext("Local tunnel IPv4 address"), gettext("Remote tunnel IPv4 address"), gettext("Remote IPv4 tunnel network"));
+	}
 
 	do_input_validation($_POST, $reqdfields, $reqdfieldsn, $input_errors);
 
-	if ((!is_ipaddr($_POST['tunnel-local-addr'])) ||
-	    (!is_ipaddr($_POST['tunnel-remote-addr'])) ||
-	    (!is_ipaddr($_POST['remote-addr']))) {
-		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IP addresses.");
+	if ($tunnel_type === '') {
+		$input_errors[] = gettext("The tunnel needs either a valid IPv4 or IPv6 tunnel configuration.");
+	}
+
+	if (((!is_ipaddr($_POST['tunnel-local-addr'])) ||
+	     (!is_ipaddr($_POST['tunnel-remote-addr']))) &&
+	    in_array($tunnel_type, array('v4', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv4 addresses.");
+	}
+
+	if (((!is_ipaddrv6($_POST['tunnel-local-addr6'])) ||
+	     (!is_ipaddrv6($_POST['tunnel-remote-addr6']))) &&
+	     in_array($tunnel_type, array('v6', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv6 addresses.");
+	}
+
+	if (!is_ipaddr($_POST['remote-addr'])) {
+		$input_errors[] = gettext("The remote address must be a valid IP address.");
 	}
 
 	if (!is_numericint($_POST['tunnel-remote-net'])) {
 		$input_errors[] = gettext("The GRE tunnel subnet must be an integer.");
 	}
 
-	if (is_ipaddrv4($_POST['tunnel-local-addr'])) {
+	if (!is_numericint($_POST['tunnel-remote-net6'])) {
+		$input_errors[] = gettext("The v6 GRE tunnel subnet must be an integer.");
+	}
+
+	if (in_array($tunnel_type, array('v4', 'v4v6'))) {
+		if (!is_ipaddrv4($_POST['tunnel-local-addr'])) {
+			$input_errors[] = gettext("The GRE Tunnel local address must be IPv4 if the tunnel type is IPv4.");
+		}
 		if (!is_ipaddrv4($_POST['tunnel-remote-addr'])) {
-			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv4 where tunnel local address is IPv4.");
+			$input_errors[] = gettext("The GRE Tunnel remote address must be IPv4 if the tunnel type is IPv4.");
```

Comment:
No need to mention the tunnel type here, the fields are separate and have distinct names, only mention specifically which field is wrong.